
# Credits

https://agupubs.onlinelibrary.wiley.com/doi/epdf/10.1029/2018GL078202



## Contents  
<br>

1. [Introduction to Neural Networks](#IntroductionToNN)

    1. [What are Neural Networks](#WhatAreNN)
    1. [Neurons](#Neurons)
    1. [Activation Functions](#ActivationFunctions)
    1. [ForwardPropagation](#ForwardPropagation)
1. [Training Neural Networks](#TrainingNN)
    1.  [Backward Propagation](#BackwardPropagation)
1. [Neural networks using Keras](#KerasNN)

# 


# Required packages

numpy
dask
xarray

pip install nbresuse


<a id='IntroductionToNN'></a>

# Introduction to Neural Networks

<a id='WhatAreNN'></a>
## What are Neural Networks

Neural networks are a type of machine learning models which are *losely inspired* in biological neurons and human nervous system. These models are used to recognize complex patterns and relationships that exists within a dataset. 

They have following properties:

1. The core architecture of a Neural Network model consists of a large number of simple processing nodes called **Neurons** which are interconnected and organized in different layers. 

2. An individual node in a layer is connected to several other nodes in the previous and the next layer. The inputs form one layer are received and processed to generate the output which is passed to the next layer.

3. The first layer of this architecture is often named as input layer which accepts the inputs, the last layer is named as the output layer which produces the output and every other layer between input and output layer is named is hidden layers. 


<img src="./fig/Neural_network_bottleneck_achitecture.svg" alt="https://upload.wikimedia.org/wikipedia/commons/8/8b/Neural_network_bottleneck_achitecture.svg" width="50%">

<a id='Neurons'></a>
## Neurons

A Neuron is a single processing unit of a Neural Network which are connected to different other neurons in the network. These connections repersents inputs and ouputs from a neuron. To each of its connections, the neuron assigns a “weight” (W) which signifies the importance the input and adds a bias (b) term. 

<img src="https://upload.wikimedia.org/wikipedia/commons/6/60/ArtificialNeuronModel_english.png" alt="https://upload.wikimedia.org/wikipedia/commons/6/60/ArtificialNeuronModel_english.png" width="85%">

<a id='ActivationFunctions'></a>
### Activation Functions 

The activation functions are used to apply non-linear transformation on input to map it to output. The aim of activation functions is to predict the right class of the target variable based on the input combination of variables. Some of the popular activation functions are Relu, Sigmoid, and TanH. 

<img src="https://cs-cheatsheet.readthedocs.io/en/latest/_images/activation_functions.png" alt="https://cs-cheatsheet.readthedocs.io/en/latest/_images/activation_functions.png" width="70%">

< Source: Stanford cs231n >

<a id='ForwardPropagation'></a>
### Forward Propagation 

Neural Network model compute the obtain the desired ouput through a process called forward propagation, in which it passes the computed activation outputs in the forward direction. 

$o_j= \varphi \left( \sum\limits_{i=1}^n x_i*w_{ij} + b_i) \right)$

Where: 

- $x_i$ : $i^{th}$ input to the neuron.
- *j* : the neuron number.
- $w_{ij}$ : weight associated with the $i^(th}$ input.
- $\varphi$ : the activation function.
- $b_i$ : the bias associated with the node for the $i^{th}$ input (neuron).
- $o_j$ : output value for the $j^{th}$ neuron.

Let's express the ouput values of a layer **O** in a more compact form using matrix notation:

$\vec{o}=\varphi \left( W \cdot \vec{x} + \vec{b} \right)$

Where: 

- $\vec{x}$ : Input vector (1D)
- $W$ : weights for each neuron and input (2D).
- $\varphi$ : the activation function.
- $\vec{b}$ : the bias associated with the node for each input (1D)
- $\vec{o}$ : output values (1D)

The arquitecture of the neural network and the weigths **W** and **b** represents our "model".
But a priori we don't know the exact value of the weigths and biases for our particular problem.

<a id='TrainingNN'></a>
## Training Neural Networks

To create a model that represents our data we adjust the weights gradually using a feedback signal from a set of know inputs and outputs, namely our training data.

To adjust our weights, we need a measure of how well our model is performing in the training data. For that we will can use a metric called "loss function" or "cost function" to measure the distance between our predicted and the expected output.

E.g.:

$Loss = Cost function = RMSE(Actual\_Values - Predicted\_Values)$
<a id='BackwardPropagation'></a>
### Backward Propagation

Neural Network model undergoes the process called backpropagation in which the error is passed to backward layers so that those layers can also improve the associated values of weights and bias. 

The backpropagation uses an algorithm called Gradient Descent in which the error is minimized and optimal values of weights and bias are obtained. This weights and bias adjustment is done by computing the derivative of error, derivative of weights, bias and subtracting them from the original values. 

<a id='KerasNN'></a>

# Neural networks using Keras

Keras is a high-level neural networks API, written in Python and capable of running on top of TensorFlow, CNTK, or Theano. It was developed with a focus on enabling fast experimentation. Being able to go from idea to result with the least possible delay is key to doing good research.


## Getting started: 30 seconds to Keras

The core data structure of Keras is a model, a way to organize layers. The simplest type of model is the Sequential model, a linear stack of layers. For more complex architectures, you should use the Keras functional API, which allows to build arbitrary graphs of layers.


Let's use keras to create a neural network like the one shown in the figure:

<img src="./fig/Neural_network_bottleneck_achitecture2.png?" alt="https://upload.wikimedia.org/wikipedia/commons/8/8b/Neural_network_bottleneck_achitecture.svg" width="50%">


Main characteristics:

- Number of input variables: **7** 
- Number of input layer nodes (1st layer): **5** 
- Number of first hidden layer nodes (2nd layer): **4**
- Number of second hidden layer nodes (3rd layer): **4**
- Number of ouput layer nodes (3th layer): **3**
- Activation functions

    - input and output layer: **relu** 
    - hidden layers: **sigmoid**

The Sequential model is a linear stack of layers.

In keras, a regular feed-forwards densely-connected NN layer is represented by the 
[keras.layers.Dense](https://keras.io/layers/core/#dense) class.

Dense implements the operation: output = activation(dot(input, kernel) + bias) where activation is the element-wise activation function passed as the activation argument, kernel is a weights matrix created by the layer, and bias is a bias vector created by the layer (only applicable if use_bias is True).


In [1]:
from keras.models import Sequential
from keras.layers import Dense

model = Sequential()

# Stacking layers is as easy as .add():

# Input layer. Here we specify the input shape.
model.add(Dense(units=5, activation='relu', input_dim=7)) 

# Hidden layers
model.add(Dense(units=4, activation='sigmoid'))
model.add(Dense(units=4, activation='sigmoid'))

#Output layer
model.add(Dense(units=3, activation='relu'))

Using TensorFlow backend.


Instructions for updating:
Colocations handled automatically by placer.


### Specifying the input shape

The model needs to know what input shape it should expect. For this reason, the first layer in a Sequential model (and only the first, because following layers can do automatic shape inference) needs to receive information about its input shape.


### Compilation

Before training a model, you need to configure the learning process, which is done via the compile method. It receives three arguments:

- An **optimizer**. This could be the string identifier of an existing optimizer (such as rmsprop or adagrad), or an instance of the Optimizer class. See: [optimizers](https://keras.io/optimizers/).
- A **loss** function. This is the objective that the model will try to minimize. It can be the string identifier of an existing loss function (such as categorical_crossentropy or mse), or it can be an objective function. See: [losses](https://keras.io/losses/).
- A list of **metrics**. For any classification problem you will want to set this to metrics=['accuracy']. A metric could be the string identifier of an existing metric or a custom metric function. [See available metrics here](https://keras.io/metrics/). The loss functions can also be used as metrics!

Let's compile our model:

In [2]:
model.compile(optimizer='sgd', loss='mse', metrics=['mse'])
# sgd = Stochastic gradient descent optimizer.             

### Training

Keras models are trained on Numpy arrays of input data and labels. For training a model, you will typically use the fit function. Read its documentation [here](https://keras.io/models/sequential/).

E.g.: Train the model, iterating on the data in batches of 32 samples

`model.fit(x_data, y_data, epochs=10, batch_size=32)`


Where: 
- **epochs**: Number of epochs to train the model. An epoch is an iteration over the entire x and y data provided. Note that in conjunction with initial_epoch, epochs is to be understood as "final epoch". The training process may consist in more than one epoch.

- **batch_size**: Number of samples per gradient update. If unspecified, batch_size will default to 32.

### Predict

The **predict** method generates output predictions for the input samples. 

`model.predict(x, batch_size=None, verbose=0, steps=None, callbacks=None)`

The computation is done in batches.


### evaluate

The **evaluate**  method returns the loss value & metrics values for the model in test mode.

`evaluate(x=None, y=None, batch_size=None, verbose=1)`

Computation is done in batches.

# Parametrize convection using Neural networks


Representing unresolved moist convection in coarse‐scale climate models remains one of the main bottlenecks of current climate simulations. The coarse resolution of the climate models is not sufficient to capture the convective processes that produce rain. In consequence, to represent the covection models rely on a parametrization that diagnose the precipitation at the gorund from the model prognostic variables (humidity, temperature, pressure, etc.).

In this part of the tutorial we will try to use a Neural Network to learn the convective parametrization used in simplified climate model. 


## The model

To generate the traning data (state variables and the precipitation) we will use the SPEEDY model.
SPEEDY is a simplified GCM developed at ICTP by Franco Molteni and Fred Kucharski.

The ICTP AGCM (nicknamed SPEEDY, for "Simplified Parameterizations, privitivE-Equation DYnamics") is based on a spectral dynamical core developed at the Geophysical Fluid Dynamics Laboratory. It is a hydrostatic, s-coordinate, spectral-transform model in the vorticity-divergence form, with semi-implicit treatment of gravity waves.

- https://www.ictp.it/research/esp/models/speedy.aspx
- Molteni F (2003) Atmospheric simulations using a GCM with simplified physical  parametrizations. I. Model climatology and variability in multi-decadal experiments. Clim Dyn 20: 175-191

- Kucharski F, Molteni F, and Bracco A (2006) Decadal interactions between the western tropical Pacific and the North Atlantic Oscillation. Clim Dyn 26: 79-91

## Training dataset

The training dataset was generated by running the model for 10 years with a spatial resolution of approx. 2 degrees. The main characteristics of the dataset are:

- T30 horizontal resoltution (approx. 2 degrees)
- Outputs variables available every 6h for a 10 yr period
- Only the latitudes between -30 and 30 degress are included

Let's take a look at the dataset.

In [3]:
import xarray as xr
import numpy as np

dataset = xr.open_dataset('gcm_run_3.nc', chunks=dict(time = 100))

Let's see the variables in the dataset.

In [4]:
list(dataset.variables.keys())

['time', 'lon', 'lat', 'lev', 'gh', 'temp', 'q', 'sp', 'precnv']

- coordinates: 'time', 'lon', 'lat', 'lev'
- prognostic variables: 'gh', 'temp', 'q', 'sp'
- diagnostic variable: 'precnv'

Let's check the times in the dataset 

In [5]:
print("First date: " , dataset['time'].values.min())
print("Last date: " , dataset['time'].values.max())

First date:  2003-01-01T00:00:00.000000000
Last date:  2003-12-31T18:00:00.000000000


Let's see in more detail the contents of the dataset.

In [6]:
dataset

<xarray.Dataset>
Dimensions:  (lat: 16, lev: 8, lon: 96, time: 1460)
Coordinates:
  * time     (time) datetime64[ns] 2003-01-01 ... 2003-12-31T18:00:00
  * lon      (lon) float64 0.0 3.75 7.5 11.25 15.0 ... 345.0 348.8 352.5 356.2
  * lat      (lat) float64 -27.83 -24.12 -20.41 -16.7 ... 16.7 20.41 24.12 27.83
  * lev      (lev) float64 925.0 850.0 700.0 500.0 300.0 200.0 100.0 30.0
Data variables:
    gh       (time, lev, lat, lon) float32 dask.array<shape=(1460, 8, 16, 96), chunksize=(100, 8, 16, 96)>
    temp     (time, lev, lat, lon) float32 dask.array<shape=(1460, 8, 16, 96), chunksize=(100, 8, 16, 96)>
    q        (time, lev, lat, lon) float32 dask.array<shape=(1460, 8, 16, 96), chunksize=(100, 8, 16, 96)>
    sp       (time, lat, lon) float32 dask.array<shape=(1460, 16, 96), chunksize=(100, 16, 96)>
    precnv   (time, lat, lon) float32 dask.array<shape=(1460, 16, 96), chunksize=(100, 16, 96)>
Attributes:
    CDI:          Climate Data Interface version 1.9.6 (http://mpimet.mpg

In nutshell, the shape of the model prognostic variables (our predictors) :

- __gh(time, lev, lat, lon)__ = **gh(1460, 8, 16, 96)** = geopotential height [m]
- __temp(time, lev, lat, lon)__ = **temp(1460, 8, 16, 96)** = abs. temperature [degK]
- __q(time, lev, lat, lon)__ = **q(1460, 8, 16, 96)** = specific humidity [g/Kg]
- __sp(time, lat, lon)__ = **sp(1460, 16, 96)** = surface pressure [hPa]

The shape of the diagnosed precipitation (the predicted variable) is:
- __precnv(time, lat, lon)__ = **precnv(1460, 16, 96)** = convective precipitation [mm/day]

In [ ]:
# Let's remove the unnecessary attributes to have cleaner outputs
dataset.attrs=dict()

## Prepare the data

We are going to predict the precipitation at each grid point using the information from the predictor variables in the corresponding column.

Therefore, the input variables to the Neural network consist of

- __Input__: $\vec{x}= \left\{ gh(z), temp(z), q(z), sp \right\}$, with a length of $\vec{x}$ is 8+8+8+1=25.
- __Output__: R, the precipitation at the ground with length=1.

The training samples in the dataset consist in many columns for every time step. Hence, based on the amount of data in dataset we have:

$total\_samples=number\_of\_times*number\_of\_lats*number\_of\_lons$

$total\_samples=1460*16*96=2242560$

**That is more than 2 millon samples per year!**


### Step 1: Collapse dimensions 

In the original dataset the dimensions of our variables are (time, lev, lat, lon) or (time, lat, lon).
What we want to feed to the Neural network are arrays with shape x=(samples, variables) and y=(samples).

Therefore, we have to collapse the (time, lat, lon) dimensions into a single dimension (sample).

We will create a function to do that for a given dataset and save the results into a file for future use.
Saving the preprocessed data to a file will also help to mainting the memory requirements to the minimum.

But, before we create the function, let's see this how this preprocessing is done in a small dataset.

In [ ]:
small_dataset = dataset.isel(time=slice(1,101))
dict(small_dataset.dims)

To collapse the (time, lat, lon) dimensions into a single dimension (sample) we will use the [Dataset's stack](http://xarray.pydata.org/en/stable/generated/xarray.Dataset.stack.html#xarray.Dataset.stack) method.

In [ ]:
stacked_dataset = small_dataset.stack(sample=('time', 'lat', 'lon'))
stacked_dataset

Let's reorder the variables dimensions as (sample, lev)

In [ ]:
stacked_dataset = stacked_dataset.transpose('sample', 'lev')
stacked_dataset

The last step is to create the x,y training samples, where their dimensions are:
- x(samples, variables)
- y(samples)

The aggregate all the column data from all the variables into a single dimension, we will use the [numpy's append](https://docs.scipy.org/doc/numpy/reference/generated/numpy.append.html) function.

In [ ]:
# The number of dimensions of x are x(samples, variables)

# "sp" dimensions were (time, lat, lon). After we collapse these dimensions into "sample"
# it became a 1D variable.

# To start appending all the column data into a the dimension variable,
# let's reshape sp to (samples, 1)
x = stacked_dataset['sp'].values[:,None]
# (None is the same as np.newaxis) 

# Now, let's continue adding the data into the second dimension
for variable in ['gh', 'temp', 'q']:
    x=np.append(x, stacked_dataset[variable].values, axis=1)

y = stacked_dataset['precnv'].values

In [ ]:
print('x.shape=', x.shape)
print('y.shape=', y.shape)

In [ ]:
new_dataset = xr.Dataset({'x': (['samples', 'variables'],  x),
                          'y': (['samples'], y)} )

new_dataset

Now let's create a function that implements the latter preprocessing step.

In [ ]:
def collapse_dims(input_dataset_path, output_dataset_path):
    """
    Prepare the training samples from the input dataset for NN feeding. 
    
    x=(samples, variables) and y=(samples).
    """
    
    _input_dataset = xr.open_dataset(input_dataset_path)
    
    _stacked_dataset = _input_dataset.stack(sample=('time', 'lat', 'lon'))
    
    _stacked_dataset = _stacked_dataset.transpose('sample', 'lev')
    
    _x = _stacked_dataset['sp'].values[:,None]
    for variable in ['gh', 'temp', 'q']:
        _x=np.append(_x, _stacked_dataset[variable].values, axis=1)

    _y = _stacked_dataset['precnv'].values
    _new_dataset = xr.Dataset({'x': (['samples', 'variables'],  _x),
                             'y': (['samples'], _y)} )
    
    
    #Save dataset
    _new_dataset.to_netcdf(output_dataset_path)
    

Now, let's process all the input files.


In [ ]:
collapse_dims('gcm_run_3.nc', 'train_3.nc')

Unfortunatelly, working with the entire dataset is not as easy, specially in computers with low ram memory.

To keep the memory requirements to the miminum, let's read our dataset as a sequence of data, that is, several (but not all) samples at a time.

For that we will make use of the Kera's [Sequence](https://keras.io/utils/#sequence) class.
The **Sequence** are a safer way to do multiprocessing. This structure guarantees that the network will only train once on each sample per epoch.

When we define a custom Sequence for our problem, the Sequence must implement the \_\_getitem\_\_ and the \_\_len\_\_ methods.
The method \_\_getitem\_\_ should return a complete batch.

### Step #: Lazy loading of all the preprocessed files

In [ ]:
input_dataset = xr.open_mfdataset(['train_3.nc'])


# i_sam = input_dataset.samples[(1<input_dataset['y']) & (input_dataset['y']<100)]

# input_dataset = input_dataset.isel(samples=i_sam)
input_dataset

In [ ]:
from sklearn.preprocessing import StandardScaler


input_scaler = StandardScaler()
input_scaler.fit(input_dataset['x'].values)

## Train a neural network


### Separate the dataset in train and validation subsets

Let's select the first 4/5 of the times for training and the last 1/5 for validation.

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.utils import shuffle

x_data = input_scaler.transform(input_dataset['x'].values)
y_data = input_dataset['y'].values

(x_train, x_test, 
 y_train, y_test) = train_test_split(x_data,
                                     y_data,
                                     test_size = 0.25,# Keep 25% of the samples for testing 
                                     shuffle=False) # Shuffle the samples

# x_train, y_train = shuffle(x_train, y_train)

In [ ]:


x_data.shape

### Construct the Neural Network

Let's construct a Neural Network with the following layers:

- Input: $\vec{x}$ (length 25)
- Output: R (length 1).
- Input layer (1st layer): **10 nodes, relu activation function** 
- 1st hidden layer (2nd layer): **10 nodes, sigmoid activation function** 
- 2nd hidden layer (3rd layer): **5 nodes, sigmoid activation function** 
- Output layer (4th layer): **1 nodes, relu activation function** 

In [ ]:
from keras.models import Sequential
from keras.layers import Dense

my_model = Sequential()
# Input layer. Here we specify the input shape.
my_model.add(Dense(units=25, activation='relu', input_dim=25)) 
# Hidden layers
my_model.add(Dense(units=25, activation='relu'))
#my_model.add(Dense(units=25, activation='relu'))
#Output layer
my_model.add(Dense(units=1, activation='relu'))

# Compile the model
my_model.compile(optimizer='Adam', loss='mse', metrics=['rmse'])


# Talk more about the options

### Train the model

Since we are using the keras Sequence to generate the data, instead of using the **fit** method, we will use 
the [fit_generator](https://keras.io/models/model/#fit_generator) method.

**fit_generator** trains the model on data generated batch-by-batch by a Python generator (or an instance of Sequence). The generator is run in parallel to the model, for efficiency. For instance, this allows you to do real-time data augmentation on images on CPU in parallel to training your model on GPU.

The use of keras.utils.Sequence guarantees the ordering and guarantees the single use of every input per epoch when using use_multiprocessing=True.

In [ ]:
# my_model.fit_generator(train_seq, epochs=1, use_multiprocessing=True, workers=2,
#                        validation_data=validation_seq, verbose=1)

In [ ]:
my_model.fit(x=x_train,y=y_train, validation_data=(x_test,y_test), epochs=20, verbose=1,shuffle=True)

In [ ]:
import gc
gc.collect()

# Credits

The description of the Keras library was adapted from the Keras official documentation at https://keras.io/.

# License

### Mit License

Copyright 2019 Andres Perez Hortal

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.